src/dune_rules/dep_rules.ml

```diff
@@ -112,21 +125,62 @@ let rec deps_of md ~ml_kind (m : Modules.Sourced_module.t) =
       | Intf -> Imported_from_vlib m
       | Impl -> Normal m)
 
+(** Tests whether a set of modules is a singleton *)
+let has_single_file modules = Option.is_some @@ Modules.as_singleton modules
+
+let immediate_deps_of unit modules obj_dir ml_kind =
+  match Module.kind unit with
+  | Alias _ -> Action_builder.return []
+  | Wrapped_compat ->
+    let interface_module =
+      match Modules.lib_interface modules with
+      | Some m -> m
+      | None -> Modules.compat_for_exn modules unit
+    in
+    List.singleton interface_module |> Action_builder.return
+  | _ ->
+    if has_single_file modules then Action_builder.return []
```

Comment:
I don't understand this case. Why is it special when there's only a single file? Can't that single file have immediate dependencies?

Reply:
That's right. If there's only a single module in the library or executable, all of its dependencies must come from libraries because a module cannot depend on itself.